Add developer instrumentation hooks / APIs for autologging (AutologgingEventLogger)

dbczumar opened the pull request:
## What changes are proposed in this pull request?

This PR introduces an `AutologgingEventLogger` developer API that provides logging hooks for important stages in the lifecycle of an autologged ML routine. It then instruments the patched functions yielded by `safe_patch` to call `AutologgingEventLogger` hooks in the appropriate locations.

## How is this patch tested?

New unit tests for event logging & existing unit / integration tests (to ensure that event logging isn't breaking existing autologging / other MLflow workflows)

## Release Notes

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

dbczumar committed:
Start

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Progress

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Instrument

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Fixes

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Progress

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Prog

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Fixes

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Black

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Rename & doc

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Some more docs

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Tests

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Merge remote-tracking branch 'origin/master' into autolog_instrument

dbczumar committed:
Format

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Nit

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/utils/autologging_utils.py:
```diff
@@ -510,17 +517,20 @@ def __call__(self, original, *args, **kwargs):
                 raise e
 
 
+AutologgingSession = namedtuple("AutologgingSession", ["integration", "id"])
```
Pull this out as a "public" developer API. Folks who supply their own `AutologgingEventLogger` implementations may have to reason about its content

dbczumar committed:
Comment

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/utils/autologging_utils.py:
```diff
@@ -533,6 +543,213 @@ def end_session(cls):
         cls._session = None
 
 
+class AutologgingEventLogger:
```
@apurva-koti @harupy This abstraction allows developers (e.g., Databricks engineers) to plug in their own event loggers in order to instrument autologging ML execution

dbczumar committed:
Move

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

apurva-koti reviewed (commented):
Some comments 🍣

harupy commented on mlflow/utils/autologging_utils.py:
```diff
@@ -326,6 +328,11 @@ def wrapper(_autolog):
         def autolog(*args, **kwargs):
+            try:
+                AutologgingEventLogger.get_logger().log_autolog_called(name, args, kwargs)
+            except Exception:  # pylint: disable=broad-except
```
+1

dbczumar replied:
+1! Will file a follow-up PR for this that adds `broad-except` to `pylintrc` and drops all inline instances

harupy commented on tests/autologging/test_autologging_safety_unit.py:
```diff
@@ -90,6 +92,95 @@ def clean_up_leaked_runs():
+
+
+@pytest.fixture
+def test_logger():
```
Can we rename `TestLogger` to `FakeLogger`, `DummyLogger`, or `MockLogger` (whatever sounds like it's a logger for testing) and rename this fixture to `fake_logger`, `dummy_logger` or `mock_logger`?

apurva-koti replied:
+1

dbczumar replied:
Sounds great! I went with `MockEventLogger`.

dbczumar committed:
Rename

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/utils/autologging_utils.py:
```diff
@@ -533,6 +546,213 @@ def end_session(cls):
+    and invoke one or more hooks (e.g., `AutologgingEventLogger.get_logger().log_autolog_called()`).
+    """
+
+    _event_logger = None
+
+    @staticmethod
+    def get_logger():
+        """
+        Fetches the configured `AutologgingEventLogger` instance for logging.
+
+        :return: The instance of `AutologgingEventLogger` specified via `set_logger`
+                 (if configured) or the default implementation of `AutologgingEventLogger`
+                 (if a logger was not configured via `set_logger`).
+        """
+        return AutologgingEventLogger._event_logger or AutologgingEventLogger()
+
+    @staticmethod
+    def set_logger(logger):
+        """
+        Configures the `AutologgingEventLogger` instance for logging. This instance
+        is exposed via `AutologgingEventLogger.get_logger()` and callers use it to invoke
+        logging hooks (e.g., AutologgingEventLogger.get_logger().log_autolog_called()).
+
+        :param logger: The instance of `AutologgingEventLogger` to use when invoking logging hooks.
+        """
+        AutologgingEventLogger._event_logger = logger
```
The goal of `AutologgingEventLogger` is to provide a way for developers (e.g., Databricks engineers working on the Machine Learning Runtime) to plug their own logging / instrumentation hooks into MLflow's autologging routines. `safe_patch` is instrumented with `AutologgingEventLogger` calls at various points in the patch execution procedure.

Accordingly, we need to tell `safe_patch` *which* hooks to use throughout the patch code path by specifying an `AutologgingEventLogger` instance (i.e. some subclass of `AutologgingEventLogger` that a developer has written). The approach taken in this PR is to use a static `set_logger()` method defined on `AutologgingEventLogger` to provide the instance to `safe_patch`, which `safe_patch` consumes via the `AutologgingEventLogger.get_logger()` method.

As an alternative, we may consider separating the hooks and the static functions into two separate objects / classes (e.g, `AutologgingEventLoggerProvider` and `AutologgingEventLogger`), though I'm not sure that it helps simplify things.

@apurva-koti Let me know if this makes things clearer, and please let me know if you have suggestions for a simpler approach. cc @harupy

dbczumar reviewed (commented):
@harupy @apurva-koti Addressed your comments! Let me know if you can take another look

dbczumar committed:
Black

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar commented on tests/autologging/test_autologging_safety_unit.py:
```diff
@@ -90,6 +92,95 @@ def clean_up_leaked_runs():
+        prev_logger = AutologgingEventLogger.get_logger()
+        logger = TestLogger()
+        AutologgingEventLogger.set_logger(logger)
+        yield logger
```
See https://github.com/mlflow/mlflow/pull/3870#discussion_r546207648 for clarification

harupy commented on mlflow/utils/autologging_utils.py:
```diff
@@ -533,6 +546,213 @@ def end_session(cls):
+        self, session, patch_obj, function_name, call_args, call_kwargs
+    ):
+        """
+        Called during the execution of a patched ML API associated with an autologging integration
```
Can we simply say `a patched API`?

```suggestion
        Called during the execution of a patched API associated with an autologging integration
```

Maybe it's just me but `ML API` sounds like this:

![Screen Shot 2020-12-21 at 12 55 36](https://user-images.githubusercontent.com/17039389/102737865-e2d42080-438b-11eb-8023-ce51981f9f48.png)

https://nordicapis.com/choosing-the-best-machine-learning-api-for-your-project/

dbczumar replied:
Sounds good! Neat link :). I've removed "ML"!

dbczumar committed:
Support universal autolog

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar committed:
Rename

Signed-off-by: Corey Zumar <corey.zumar@databricks.com>

dbczumar commented:
@harupy Thanks for the additional feedback. I've addressed the naming in docstrings (`ML API` -> `API`) and also added support / test coverage for event logging from calls to `mlflow.autolog()` (by adding the `autologging_integration` decorator with name `mlflow` to `mlflow.autolog()`). Let me know how things look!

harupy reviewed (approved):
LGTM :)

dbczumar commented:
> LGTM :)

Thank you!
